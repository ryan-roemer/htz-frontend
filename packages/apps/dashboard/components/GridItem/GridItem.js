import React from 'react';
import { FelaComponent, } from 'react-fela';
import { parseComponentProp, parseStyleProps, } from '@haaretz/htz-css-tools';

// ////////////////////////////////////////////////////////////////// //
//                             PROP-TYPES                             //
// ////////////////////////////////////////////////////////////////// //

// ////////////////////////////////////////////////////////////////// //
//                               STYLES                               //
// ////////////////////////////////////////////////////////////////// //
const gridItemStyles = ({
                          gutter,
                          offset,
                          gridRule: rule,
                          stretchContent,
                          width,
                          miscStyles,
                          theme,
                        }) => ({
  // Include gutter width (padding) in total width calculation
  boxSizing: 'border-box',
  // Allow `<GridItem>`s without a set width to auto-fill the row they are in
  // eslint-disable-next-line eqeqeq
  flexBasis: width == null ? 0 : 'auto',
  flexGrow: 1,
  flexShrink: 1,
  listStyle: 'none',
  paddingInlineEnd: `${gutter / 2}rem`,
  paddingInlineStart: `${gutter / 2}rem`,

  extend: [
    ...(stretchContent
      ? [
        parseComponentProp(
          'stretchContent',
          stretchContent,
          theme.mq,
          contentStretcher
        ),
      ]
      : []),
    // Offset an item from the previous item (or the begining of the grid)
    ...(offset
      ? [ parseComponentProp('offset', offset, theme.mq, setOffset), ]
      : []),
    // Set the vertical rule at the end of the `<GridItem />`
    ...(rule
      ? [ parseComponentProp('rule', rule, theme.mq, setRule, theme), ]
      : []),
    // Set the width of an item
    ...(width ? [ parseComponentProp('width', width, theme.mq, setWidth), ] : []),
    // Trump all other styles with those defined in `miscStyles`
    ...(miscStyles ? parseStyleProps(miscStyles, theme.mq, theme.type) : []),
  ],
});

// ///////////////// //
//  Styling Helpers  //
// ///////////////// //
function contentStretcher(prop, stretchContent) {
  return stretchContent
    ? {
      display: 'flex',
      flexDirection: 'column',
    }
    : {};
}

function setOffset(prop, offset) {
  const offsetInPercent = `${offset * 100}%`;
  return {
    marginInlineStart: offsetInPercent,
  };
}

function setWidth(prop, width) {
  const widthInRem = `${width}rem`;
  const widthInPercent = `${width * 100}%`;
  const widthToUse = width > 1
    ? widthInRem // fixed width in rems
    : widthInPercent; // relative width in percentage
  return {
    // Keep `<GridItem>`s at their prescribed size
    flexGrow: 0,
    flexShrink: 0,
    flexBasis: widthToUse,
    maxWidth: widthToUse,
  };
}

function setRule(prop, ruleOptions, theme) {
  const ruleWidth = ruleOptions.width || theme.gridStyle.ruleWidth;
  const ruleColor = ruleOptions.color || theme.gridStyle.ruleColor;
  const ruleColorAsArray = Array.isArray(ruleColor) ? ruleColor : [ ruleColor, ];
  return {
    position: 'relative',

    ':after': {
      backgroundColor: theme.color(...ruleColorAsArray),
      content: '""',
      height: '100%',
      // This depends on `fela-plugin-bidi`
      insetInlineEnd: '0',
      position: 'absolute',
      top: '0',
      // This depends on `fela-plugin-bidi`
      // We set the rule at the half its width past the gutter's end,
      // so it is placed exactly between two `<GridItem>`s. However, we
      // round floats into integers to avoid fuzzy rendering of subpixels.
      transform: `logical translateX(${Math.floor(ruleWidth / 2)}px)`,
      width: `${ruleWidth}px`,
    },
  };
}

GridItem.defaultProps = {
  attrs: null,
  children: null,
  id: null,
  gutter: null,
  offset: null,
  rule: null,
  stretchContent: false,
  tagName: 'div',
  width: null,
  miscStyles: null,
};

// ////////////////////////////////////////////////////////////////// //
//                             COMPONENT                              //
// ////////////////////////////////////////////////////////////////// //
export default function GridItem({
  children,
  gutter,
  id,
  offset,
  rule,
  stretchContent,
  tagName,
  width,
  attrs,
  miscStyles,
}) {
  return (
    <FelaComponent
      {...{
        gutter,
        offset,
        gridRule: rule,
        stretchContent,
        width,
        miscStyles,
      }}
      rule={gridItemStyles}
      render={({ className, }) => {
        const GridItemElement = tagName;
        return (
          <GridItemElement id={id} className={className} {...attrs}>
            {children}
          </GridItemElement>
        );
      }}
    />
  );
}
