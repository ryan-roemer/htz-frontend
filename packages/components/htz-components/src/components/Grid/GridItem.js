import React from 'react';
import PropTypes from 'prop-types';
import { createComponent, } from 'react-fela';
import { parseComponentProp, parseStyleProps, } from '@haaretz/htz-css-tools';
import { attrsPropType, } from '../../propTypes/attrsPropType';
import { responsivePropBaseType, } from '../../propTypes/responsivePropBaseType';
import { stylesPropType, } from '../../propTypes/stylesPropType';

/**
 * Properties of a `<GridItem />`
 */
const StyledGridItemPropTypes = {
  // Underlying component props //
  /**
   * An object of attributes to set on the DOM element.
   */
  attrs: attrsPropType,
  /**
   * Nodes rendered inside `<GridItem />`.
   */
  children: PropTypes.node,
  /**
   * The `<GridItem />`s `id` attribute.
   */
  id: PropTypes.string,
  /** The HTML tag to render the `<GridItem />` as */
  tagName: PropTypes.string,

  // Styling props //
  /**
   * The width (in `rem`) of the gutter between `<GridItem />`s in a `<Grid />`.
   * The parent `<Grid />` component will automatically augment its children
   * with this prop, so it should never be set directly on the component itself.
   */
  gutter: PropTypes.number,
  /**
   * The amount a `<GridItem />` is offs by.
   * The number passed should be (`offset` / 'columns').
   * Can be set responsively.
   *
   * @example
   * // Offset <GridItem /> by 25% (3 of 12 columns)
   * <GridItem offset={3 / 12} />
   */
  offset: PropTypes.oneOfType([
    PropTypes.number,
    PropTypes.arrayOf(
      PropTypes.shape({
        ...responsivePropBaseType,
        value: PropTypes.number.isRequired,
      })
    ),
  ]),
  /**
   * The color and width (in pixels) of a vertical devider rule placed at the end of
   * the `<GridItem />`'s gutter.
   * Can be set responsively.
   */
  rule: PropTypes.oneOfType([
    PropTypes.bool,
    PropTypes.shape({
      width: PropTypes.number.isRequired,
      color: PropTypes.oneOfType([
        PropTypes.string,
        PropTypes.arrayOf(PropTypes.string),
      ]).isRequired,
    }),
    PropTypes.arrayOf(
      PropTypes.shape({
        ...responsivePropBaseType,
        value: PropTypes.oneOfType([
          PropTypes.bool,
          PropTypes.shape({
            width: PropTypes.number.isRequired,
            color: PropTypes.oneOfType([
              PropTypes.string,
              PropTypes.arrayOf(PropTypes.string),
            ]).isRequired,
          }),
        ]).isRequired,
      })
    ),
  ]),
  /**
   * The width of a `<gridItem />`.
   * The number passed should be (`width` / `columns`).
   * When the number passed to `width` is greater than `1`, it will be
   * used as an absolute width in rems.
   *
   * Can be set responsively.
   *
   * @example
   * // <GridItem /> spans 25% (3 of 12 columns)
   * <GridItem width={3 / 12} />
   */
  width: PropTypes.oneOfType([
    PropTypes.number,
    PropTypes.arrayOf(
      PropTypes.shape({
        ...responsivePropBaseType,
        value: PropTypes.number.isRequired,
      })
    ),
  ]),
  /**
   * A special property holding miscellaneous CSS values that
   * trump all default values. Processed by
   * [`parseStyleProps`](https://Haaretz.github.io/htz-frontend/htz-css-tools#parsestyleprops)
   */
  miscStyles: stylesPropType,
};

/** The default values of a `<GridItem>`'s props */
const StyledGridItemDefaultProps = {
  attrs: null,
  children: null,
  id: null,
  gutter: null,
  offset: null,
  rule: null,
  tagName: 'div',
  width: null,
  miscStyles: null,
};

const gridItemStyles = ({
  gutter,
  offset,
  rule,
  width,
  miscStyles,
  theme,
  ...props
}) => ({
  // Include gutter width (padding) in total width calculation
  boxSizing: 'border-box',
  // Allow `<GridItem>`s without a set width to auto-fill the row they are in
  flexBasis: 'auto',
  flexGrow: 1,
  flexShrink: 1,
  listStyle: 'none',
  paddingLeft: `${gutter / 2}rem`,
  paddingRight: `${gutter / 2}rem`,

  extend: [
    // Offset an item from the previous item (or the begining of the grid)
    ...(offset
      ? [ parseComponentProp('offset', offset, theme.mq, setOffset), ]
      : []),
    // Set the vertical rule at the end of the `<GridItem />`
    ...(rule
      ? [ parseComponentProp('rule', rule, theme.mq, setRule, theme), ]
      : []),
    // Set the width of an item
    ...(width ? [ parseComponentProp('width', width, theme.mq, setWidth), ] : []),
    // Trump all other styles with those defined in `miscStyles`
    ...(miscStyles ? parseStyleProps(miscStyles, theme.mq, theme.type) : []),
  ],
});

function setOffset(prop, offset) {
  const offsetInPercent = `${offset * 100}%`;
  return {
    marginInlineStart: offsetInPercent,
  };
}

function setWidth(prop, width) {
  const widthInRem = `${width}rem`;
  const widthInPercent = `${width * 100}%`;
  return {
    // Keep `<GridItem>`s at their prescribed size
    flexGrow: 0,
    flexShrink: 0,
    flexBasis:
      width > 1
        ? // fixed width in rems
        widthInRem
        : // relative width in percentage
        widthInPercent,
  };
}

function setRule(prop, ruleOptions, theme) {
  const ruleWidth = ruleOptions.width || theme.gridStyle.ruleWidth;
  const ruleColor = ruleOptions.color || theme.gridStyle.ruleColor;
  const ruleColorAsArray = Array.isArray(ruleColor) ? ruleColor : [ ruleColor, ];
  return {
    position: 'relative',

    ':after': {
      backgroundColor: theme.color(...ruleColorAsArray),
      content: '""',
      height: '100%',
      // This depends on `fela-plugin-bidi`
      insetInlineEnd: '0',
      position: 'absolute',
      top: '0',
      // This depends on `fela-plugin-bidi`
      // We set the rule at the half its width past the gutter's end,
      // so it is placed exactly between two `<GridItem>`s. However, we
      // round floats into integers to avoid fuzzy rendering of subpixels.
      transform: `logical translateX(${Math.floor(ruleWidth / 2)}px)`,
      width: `${ruleWidth}px`,
    },
  };
}

GridItem.propTypes = StyledGridItemPropTypes;
GridItem.defaultProps = StyledGridItemDefaultProps;

/**
 * A generic, stylable GridItem component
 * @param {Object} props
 */
export function GridItem({
  attrs,
  className, // eslint-disable-line react/prop-types
  children,
  id,
  tagName,
  ...props
}) {
  const GridItemElement = tagName;
  return (
    <GridItemElement id={id} className={className} {...attrs}>
      {children}
    </GridItemElement>
  );
}

const StyledGridItem = createComponent(gridItemStyles, GridItem, [
  'attrs',
  'tagName',
]);

StyledGridItem.propTypes = StyledGridItemPropTypes;
StyledGridItem.defaultProps = StyledGridItemDefaultProps;

export default StyledGridItem;
