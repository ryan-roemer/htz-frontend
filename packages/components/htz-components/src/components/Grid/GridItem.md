<!-- START doctoc generated TOC please keep comment here to allow auto update -->
<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->
**Table of Contents**

- [**Basic Usage**](#basic-usage)
- [**Offset**](#offset)
- [**Rule**](#rule)
- [**DOM Element Attributes**](#dom-element-attributes)
- [**Custom Tag**](#custom-tag)

<!-- END doctoc generated TOC please keep comment here to allow auto update -->

`<GridItem />`s may only be placed as the direct descendants of a grid container.

Each `<GridItem>` has padding the length of half a single gutter at its left and 
right edges, creating a gap between the different items. The gutter's width is 
modified on the grid level, not that on the individual `<GridItem>`

`<GridItem>`s are set to box-sizing: border-box, so that the padding used to 
form gutters is taken into consideration when setting item widths.

#### **Basic Usage**

A `<gridItem>`'s width is set using the `width` prop, which takes a number.

When `width` is greater than `0` and less then or equal to `1`, it represents
(`width of item` / 'columns in grid row'), and, for readability,
should also be passed this way, e.g., `width={3 / 4}`. 

`width` can be set responsively (see example below).

```jsx
const fpo = {
  backgroundColor: '#0B7EB5',
  color: '#fff',
  height: '4rem',
  marginBottom: '2rem',
  padding: '1rem',
  width: '100%',
};

<Grid>
  <GridItem width={1/1}><div style={fpo} /></GridItem>
  <GridItem width={1/12}><div style={fpo} /></GridItem>
  <GridItem width={11/12}><div style={fpo} /></GridItem>
  <GridItem width={2/12}><div style={fpo} /></GridItem>
  <GridItem width={10/12}><div style={fpo} /></GridItem>
  <GridItem width={3/12}><div style={fpo} /></GridItem>
  <GridItem width={9/12}><div style={fpo} /></GridItem>
  <GridItem width={4/12}><div style={fpo} /></GridItem>
  <GridItem width={8/12}><div style={fpo} /></GridItem>
  <GridItem width={5/12}><div style={fpo} /></GridItem>
  <GridItem width={7/12}><div style={fpo} /></GridItem>
  <GridItem width={6/12}><div style={fpo} /></GridItem>
  <GridItem width={6/12}><div style={fpo} /></GridItem>

  <GridItem width={7/12}><div style={fpo} /></GridItem>
  <GridItem width={5/12}><div style={fpo} /></GridItem>
  <GridItem width={8/12}><div style={fpo} /></GridItem>
  <GridItem width={4/12}><div style={fpo} /></GridItem>
  <GridItem width={9/12}><div style={fpo} /></GridItem>
  <GridItem width={3/12}><div style={fpo} /></GridItem>
  <GridItem width={10/12}><div style={fpo} /></GridItem>
  <GridItem width={2/12}><div style={fpo} /></GridItem>
  <GridItem width={11/12}><div style={fpo} /></GridItem>
  <GridItem width={1/12}><div style={fpo} /></GridItem>
  <GridItem width={1/1}><div style={fpo} /></GridItem>
</Grid>
```

When `width` is a number greater than `1`, it will be used to set a fixed width
in `rem`s, thus supporting fixed-fluid layouts.

`<GridItem>`s lacking the `width` prop will be equally distributed in the
remaining space.

```jsx
const fpo = {
  backgroundColor: '#0B7EB5',
  color: '#fff',
  marginBottom: '2rem',
  padding: '1rem',
  textAlign: 'center',
  width: '100%',
};

<Grid>
  <GridItem width={40}><div style={fpo}>fixed to 40rem</div></GridItem>
  <GridItem><div style={fpo}>distributed in remaining space</div></GridItem>
  <GridItem><div style={fpo}>distributed in remaining space</div></GridItem>
</Grid>
```


<a name="responsiveWidth"> </a>

As previously mentioned, A `<GridItem />`'s width, may be set 
responsively by passing an array of objects specifying meqia-query 
conditions (`from`, `until`, `misc`, `type`) and a `value`, e.g.:

``` jsx static
<GridItem
   width={[ { until: 'xl', value: 2 / 4, }, { from: 'xl', value: 1 / 1, }, ]}
 > 
   ...
 </GridItem>
```

The following example breaks into two rows in viewports smaller than the `xl` 
breakpoint

```jsx
const fpo = {
  backgroundColor: '#0B7EB5',
  color: '#fff',
  height: '8rem',
  marginBottom: '4rem',
  width: '100%',
};

<Grid>
  <GridItem width={[ { until: 'xl', value: 1 / 2, }, { from: 'xl', value: 1 / 4, }, ]} >
    <div style={fpo} />
  </GridItem>
  <GridItem width={[ { until: 'xl', value: 1 / 2, }, { from: 'xl', value: 3 / 4, }, ]} >
    <div style={fpo} />
  </GridItem>
  <GridItem width={[ { until: 'xl', value: 1 / 2, }, { from: 'xl', value: 3 / 4, }, ]} >
    <div style={fpo} />
  </GridItem>
  <GridItem width={[ { until: 'xl', value: 1 / 2, }, { from: 'xl', value: 1 / 4, }, ]}>
    <div style={fpo} />
  </GridItem>
</Grid>
```

#### **Offset**

`<GridItem>`s may be pushed away from their `<Grid>`'s start by using the `offset`
prop, which work in a similar manner to the `width` prop, taking the same 
argument structure, but sets a margin to push the `<GridItem>` away instead of 
setting its width. E.g. `offset={2 / 8}` will offset the `<GridItem>` away from
it natural position by 25% of the `<Grid>`'s width (equal to two of eight 
columns).

A `<gridItem>`'s offset can be set responsively by passing an array of 
objects specifying meqia-query conditions (`from`, `until`, `misc`, `type`) 
and a `value`. See <code>[width](#responsiveWidth)</code> for a code
example.

```jsx
const fpo = {
  backgroundColor: '#0B7EB5',
  color: '#fff',
  height: '4rem',
  marginBottom: '2rem',
  padding: '1rem',
  width: '100%',
};

<Grid attrs={{ dir: 'rtl' }}>
  <GridItem offset={5/6} width={1/6}><div style={fpo} /></GridItem>
  <GridItem width={1/1} />
  <GridItem offset={4/6} width={1/6}><div style={fpo} /></GridItem>
  <GridItem width={1/1} />
  <GridItem offset={3/6} width={1/6}><div style={fpo} /></GridItem>
  <GridItem width={1/1} />
  <GridItem offset={2/6} width={1/6}><div style={fpo} /></GridItem>
  <GridItem width={1/1} />
  <GridItem offset={1/6} width={1/6}><div style={fpo} /></GridItem>
  <GridItem width={1/1} />
  <GridItem width={1/6}><div style={fpo} /></GridItem>
</Grid>
```


#### **Rule**

Using the `rule` prop, `<GridItem>`s may be appended a gutter-rule of varying 
width and color to visually separate them from the following `<GridItem>`.

The prop takes either a `boolean`, in which case the default rule settings
set in the theme will be used, or an object with two keys: 

* `color` (`string`|`string[]`), which should point to a named color from the 
theme
* `width` (`number`), which determines the rule's width in _pixels_

A `<gridItem>`'s `rule` can be set responsively by passing an array of 
objects specifying meqia-query conditions (`from`, `until`, `misc`, `type`) 
and a `value` (`boolean`|`Object`), e.g.:
``` js static
rule={[ { from: 'xl', value: true }, ]}
// or
rule={[ { from: 'xl', value: { color: [ 'neutral', '-3', ], width: 3, }, }, ]}
```

See <code>[width](#responsiveWidth)</code> for a live example of using 
responsive values.

<small><em>(The grid below is in "rtl")</em></small>

```jsx
const fpo = {
  backgroundColor: '#0B7EB5',
  color: '#fff',
  padding: '1rem',
  textAlign: 'center',
  width: '100%',
};

<Grid attrs={{ dir: 'rtl' }}>
  <GridItem width={1/6}><div style={fpo}>No rule</div></GridItem>
  <GridItem width={1/6}><div style={fpo}>No rule</div></GridItem>
  <GridItem width={2/6} rule><div style={fpo}>Has rule</div></GridItem>
  <GridItem width={1/6}><div style={fpo}>No rule</div></GridItem>
  <GridItem width={1/6}><div style={fpo}>No rule</div></GridItem>
</Grid>
```



#### **DOM Element Attributes**

The `attrs` prop take an object which is spread on the DOM element, and can
be used to add additional attributes to the DOM element, e.g.:

```jsx static
<GridItem attrs={{ 'aria-hidden': true, tabIndex: -1, }}> ... </GridItem>
```


```jsx
const fpo = {
  backgroundColor: '#0B7EB5',
  color: '#fff',
  height: '4rem',
  marginBottom: '2rem',
  padding: '1rem',
  width: '100%',
};

<Grid>
  <GridItem width={1/2} attrs={{ 'aria-hidden': true, tabIndex: -1, }}><div style={fpo} /></GridItem>
  <GridItem width={1/2}><div style={fpo} /></GridItem>
</Grid>
```

#### **Custom Tag**

By default, a `<GridItem />` is rendered as a div, but you are free to render is as whatever tag 
you'd like to fit the semantics of your use-case by passing a `string` to the `tagName` prop,
e.g.:

```jsx static
<GridItem tagName="section"> ... </GridItem>
```

```jsx
const fpo = {
  backgroundColor: '#0B7EB5',
  color: '#fff',
  height: '4rem',
  marginBottom: '2rem',
  padding: '1rem',
  width: '100%',
};

<Grid>
  <GridItem width={1/2} tagName="section"><div style={fpo} /></GridItem>
  <GridItem width={1/2} tagName="section"><div style={fpo} /></GridItem>
</Grid>
```
