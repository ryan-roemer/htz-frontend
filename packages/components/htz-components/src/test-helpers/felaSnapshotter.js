import React from 'react';
import createRenderer from '../components/StyleProvider/createRenderer';
import StyleProvider from '../components/StyleProvider/StyleProvider';
import createFelaSnapshotter from './createFelaSnapshotter';

const testRenderer = createRenderer({ isRtl: true, });
const felaSnapshotter = createFelaSnapshotter(StyleProvider, testRenderer);

export default felaSnapshotter;
