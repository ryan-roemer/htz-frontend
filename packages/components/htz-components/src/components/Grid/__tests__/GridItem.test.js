/* global window */
/* eslint-disable import/no-named-as-default */
import React from 'react';
import felaSnapshotter from '../../../test-helpers/felaSnapshotter';
import GridItem from '../GridItem';

describe('<GridItem />', () => {
  it('spread the "attrs" prop on the Dom element', () => {
    const { component, } = felaSnapshotter(
      <GridItem
        attrs={{
          tabIndex: '-1',
        }}
      />
    );

    expect(component).toMatchSnapshot();
  });

  it('set the "id" prop on the Dom element', () => {
    const { component, } = felaSnapshotter(<GridItem id="hasId" />);

    expect(component).toMatchSnapshot();
  });

  it('change the DOM element\'s tagName based on the "tagName" prop', () => {
    const { component, } = felaSnapshotter(<GridItem tagName="article" />);

    expect(component.children[0].type).toBe('article');
  });

  describe('styles', () => {
    it('create correct default styels', () => {
      const { component, styles, } = felaSnapshotter(<GridItem />);

      expect(component).toMatchSnapshot();
      expect(
        styles.replace(/(\.|@)/g, (match, g1) => `\n${g1}`)
      ).toMatchSnapshot();
    });

    it('apply correct gutter styles', () => {
      const { component, styles, } = felaSnapshotter(<GridItem gutter={8} />);

      expect(component).toMatchSnapshot();
      expect(
        styles.replace(/(\.|@)/g, (match, g1) => `\n${g1}`)
      ).toMatchSnapshot();
    });

    it('apply correct offset styles', () => {
      const { component, styles, } = felaSnapshotter(
        <GridItem offset={1 / 4} />
      );

      expect(component).toMatchSnapshot();
      expect(
        styles.replace(/(\.|@)/g, (match, g1) => `\n${g1}`)
      ).toMatchSnapshot();
    });

    it('apply correct default rule styles', () => {
      const { component, styles, } = felaSnapshotter(<GridItem rule />);

      expect(component).toMatchSnapshot();
      expect(
        styles.replace(/(\.|@)/g, (match, g1) => `\n${g1}`)
      ).toMatchSnapshot();
    });

    it('apply correct rule styles when ruleWidth is even', () => {
      const { component, styles, } = felaSnapshotter(
        <GridItem rule={{ width: 2, color: [ 'primary', 'base', ], }} />
      );

      expect(component).toMatchSnapshot();
      expect(
        styles.replace(/(\.|@)/g, (match, g1) => `\n${g1}`)
      ).toMatchSnapshot();
    });

    it('apply correct rule styles when ruleWidth is odd', () => {
      const { component, styles, } = felaSnapshotter(
        <GridItem rule={{ width: 5, color: [ 'primary', 'base', ], }} />
      );

      expect(component).toMatchSnapshot();
      expect(
        styles.replace(/(\.|@)/g, (match, g1) => `\n${g1}`)
      ).toMatchSnapshot();
    });

    it('apply correct rule styles when rule color is a string', () => {
      const { component, styles, } = felaSnapshotter(
        <GridItem rule={{ width: 2, color: 'black', }} />
      );

      expect(component).toMatchSnapshot();
      expect(
        styles.replace(/(\.|@)/g, (match, g1) => `\n${g1}`)
      ).toMatchSnapshot();
    });

    it('apply correct width styles (relative width)', () => {
      const { component, styles, } = felaSnapshotter(<GridItem width={1 / 4} />);

      expect(component).toMatchSnapshot();
      expect(
        styles.replace(/(\.|@)/g, (match, g1) => `\n${g1}`)
      ).toMatchSnapshot();
    });

    it('apply correct width styles (fixed width)', () => {
      const { component, styles, } = felaSnapshotter(<GridItem width={20} />);

      expect(component).toMatchSnapshot();
      expect(
        styles.replace(/(\.|@)/g, (match, g1) => `\n${g1}`)
      ).toMatchSnapshot();
    });

    it('set misc styles correctly', () => {
      const { component, styles, } = felaSnapshotter(
        <GridItem miscStyles={{ backgroundColor: 'red', }} />
      );

      expect(component).toMatchSnapshot();
      expect(
        styles.replace(/(\.|@)/g, (match, g1) => `\n${g1}`)
      ).toMatchSnapshot();
    });
  });
});
