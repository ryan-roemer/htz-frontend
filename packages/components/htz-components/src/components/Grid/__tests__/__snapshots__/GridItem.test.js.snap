// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GridItem /> set the "id" prop on the Dom element 1`] = `
<div>
  <div
    className="htz-a htz-b htz-c htz-d htz-e htz-f htz-g htz-h"
    id="hasId"
  />
</div>
`;

exports[`<GridItem /> spread the "attrs" prop on the Dom element 1`] = `
<div>
  <div
    className="htz-a htz-b htz-c htz-d htz-e htz-f htz-g htz-h"
    id={null}
    tabIndex="-1"
  />
</div>
`;

exports[`<GridItem /> styles apply correct default rule styles 1`] = `
<div>
  <div
    className="htz-a htz-b htz-c htz-d htz-e htz-f htz-g htz-h htz-i htz-j htz-k htz-l htz-m htz-n htz-o htz-p htz-q"
    id={null}
  />
</div>
`;

exports[`<GridItem /> styles apply correct default rule styles 2`] = `
"
.htz-a{box-sizing:border-box}
.htz-b{flex-basis:0}
.htz-c{flex-grow:1}
.htz-d{flex-shrink:1}
.htz-e{list-style:none}
.htz-f{padding-left:0rem}
.htz-g{padding-right:0rem}
.htz-h{position:relative}
.htz-i:after{background-color:#CCC}
.htz-j:after{content:\\"\\"}
.htz-k:after{height:100%}
.htz-l:after{left:0}
.htz-m:after{position:absolute}
.htz-n:after{top:0}
.htz-o:after{transform:translateX(0px)}
.htz-p:after{width:1px}
.htz-q{-webkit-box-flex:1}"
`;

exports[`<GridItem /> styles apply correct gutter styles 1`] = `
<div>
  <div
    className="htz-a htz-b htz-c htz-d htz-e htz-f htz-g htz-h"
    id={null}
  />
</div>
`;

exports[`<GridItem /> styles apply correct gutter styles 2`] = `
"
.htz-a{box-sizing:border-box}
.htz-b{flex-basis:0}
.htz-c{flex-grow:1}
.htz-d{flex-shrink:1}
.htz-e{list-style:none}
.htz-f{padding-left:4rem}
.htz-g{padding-right:4rem}
.htz-h{-webkit-box-flex:1}"
`;

exports[`<GridItem /> styles apply correct offset styles 1`] = `
<div>
  <div
    className="htz-a htz-b htz-c htz-d htz-e htz-f htz-g htz-h htz-i"
    id={null}
  />
</div>
`;

exports[`<GridItem /> styles apply correct offset styles 2`] = `
"
.htz-a{box-sizing:border-box}
.htz-b{flex-basis:0}
.htz-c{flex-grow:1}
.htz-d{flex-shrink:1}
.htz-e{list-style:none}
.htz-f{padding-left:0rem}
.htz-g{padding-right:0rem}
.htz-h{margin-right:25%}
.htz-i{-webkit-box-flex:1}"
`;

exports[`<GridItem /> styles apply correct rule styles when rule color is a string 1`] = `
<div>
  <div
    className="htz-a htz-b htz-c htz-d htz-e htz-f htz-g htz-h htz-i htz-j htz-k htz-l htz-m htz-n htz-o htz-p htz-q"
    id={null}
  />
</div>
`;

exports[`<GridItem /> styles apply correct rule styles when rule color is a string 2`] = `
"
.htz-a{box-sizing:border-box}
.htz-b{flex-basis:0}
.htz-c{flex-grow:1}
.htz-d{flex-shrink:1}
.htz-e{list-style:none}
.htz-f{padding-left:0rem}
.htz-g{padding-right:0rem}
.htz-h{position:relative}
.htz-i:after{background-color:#000}
.htz-j:after{content:\\"\\"}
.htz-k:after{height:100%}
.htz-l:after{left:0}
.htz-m:after{position:absolute}
.htz-n:after{top:0}
.htz-o:after{transform:translateX(-1px)}
.htz-p:after{width:2px}
.htz-q{-webkit-box-flex:1}"
`;

exports[`<GridItem /> styles apply correct rule styles when ruleWidth is even 1`] = `
<div>
  <div
    className="htz-a htz-b htz-c htz-d htz-e htz-f htz-g htz-h htz-i htz-j htz-k htz-l htz-m htz-n htz-o htz-p htz-q"
    id={null}
  />
</div>
`;

exports[`<GridItem /> styles apply correct rule styles when ruleWidth is even 2`] = `
"
.htz-a{box-sizing:border-box}
.htz-b{flex-basis:0}
.htz-c{flex-grow:1}
.htz-d{flex-shrink:1}
.htz-e{list-style:none}
.htz-f{padding-left:0rem}
.htz-g{padding-right:0rem}
.htz-h{position:relative}
.htz-i:after{background-color:#0B7EB5}
.htz-j:after{content:\\"\\"}
.htz-k:after{height:100%}
.htz-l:after{left:0}
.htz-m:after{position:absolute}
.htz-n:after{top:0}
.htz-o:after{transform:translateX(-1px)}
.htz-p:after{width:2px}
.htz-q{-webkit-box-flex:1}"
`;

exports[`<GridItem /> styles apply correct rule styles when ruleWidth is odd 1`] = `
<div>
  <div
    className="htz-a htz-b htz-c htz-d htz-e htz-f htz-g htz-h htz-i htz-j htz-k htz-l htz-m htz-n htz-o htz-p htz-q"
    id={null}
  />
</div>
`;

exports[`<GridItem /> styles apply correct rule styles when ruleWidth is odd 2`] = `
"
.htz-a{box-sizing:border-box}
.htz-b{flex-basis:0}
.htz-c{flex-grow:1}
.htz-d{flex-shrink:1}
.htz-e{list-style:none}
.htz-f{padding-left:0rem}
.htz-g{padding-right:0rem}
.htz-h{position:relative}
.htz-i:after{background-color:#0B7EB5}
.htz-j:after{content:\\"\\"}
.htz-k:after{height:100%}
.htz-l:after{left:0}
.htz-m:after{position:absolute}
.htz-n:after{top:0}
.htz-o:after{transform:translateX(-2px)}
.htz-p:after{width:5px}
.htz-q{-webkit-box-flex:1}"
`;

exports[`<GridItem /> styles apply correct width styles (fixed width) 1`] = `
<div>
  <div
    className="htz-a htz-b htz-c htz-d htz-e htz-f htz-g htz-h"
    id={null}
  />
</div>
`;

exports[`<GridItem /> styles apply correct width styles (fixed width) 2`] = `
"
.htz-a{box-sizing:border-box}
.htz-b{flex-basis:20rem}
.htz-c{flex-grow:0}
.htz-d{flex-shrink:0}
.htz-e{list-style:none}
.htz-f{padding-left:0rem}
.htz-g{padding-right:0rem}
.htz-h{-webkit-box-flex:0}"
`;

exports[`<GridItem /> styles apply correct width styles (relative width) 1`] = `
<div>
  <div
    className="htz-a htz-b htz-c htz-d htz-e htz-f htz-g htz-h"
    id={null}
  />
</div>
`;

exports[`<GridItem /> styles apply correct width styles (relative width) 2`] = `
"
.htz-a{box-sizing:border-box}
.htz-b{flex-basis:25%}
.htz-c{flex-grow:0}
.htz-d{flex-shrink:0}
.htz-e{list-style:none}
.htz-f{padding-left:0rem}
.htz-g{padding-right:0rem}
.htz-h{-webkit-box-flex:0}"
`;

exports[`<GridItem /> styles create correct default styels 1`] = `
<div>
  <div
    className="htz-a htz-b htz-c htz-d htz-e htz-f htz-g htz-h"
    id={null}
  />
</div>
`;

exports[`<GridItem /> styles create correct default styels 2`] = `
"
.htz-a{box-sizing:border-box}
.htz-b{flex-basis:0}
.htz-c{flex-grow:1}
.htz-d{flex-shrink:1}
.htz-e{list-style:none}
.htz-f{padding-left:0rem}
.htz-g{padding-right:0rem}
.htz-h{-webkit-box-flex:1}"
`;

exports[`<GridItem /> styles set misc styles correctly 1`] = `
<div>
  <div
    className="htz-a htz-b htz-c htz-d htz-e htz-f htz-g htz-h htz-i"
    id={null}
  />
</div>
`;

exports[`<GridItem /> styles set misc styles correctly 2`] = `
"
.htz-a{box-sizing:border-box}
.htz-b{flex-basis:0}
.htz-c{flex-grow:1}
.htz-d{flex-shrink:1}
.htz-e{list-style:none}
.htz-f{padding-left:0rem}
.htz-g{padding-right:0rem}
.htz-h{background-color:red}
.htz-i{-webkit-box-flex:1}"
`;
